### PR TITLE
some small code cleanup

### PR DESCRIPTION
**Pull request by CamDavidsonPilon**
This is some small cleanup of code. The biggest change is moving `WinStayLoseShift` to be a `MemoryOnePlayer` subclass. This offers less code duplication. Thoughts?

### CODE REVIEW

File: axelrod/strategies/memoryone.py [hunk 1]
```diff
@@ -71,21 +37,33 @@ def __init__(self, four_vector=None, initial='C'):
...
+            raise ValueError('An element in the probability vector, %s, is not between 0 and 1.' % four_vector)
+
         self._four_vector = dict(zip([('C', 'C'), ('C', 'D'), ('D', 'C'), ('D', 'D')], map(float, four_vector)))
         self.classifier['stochastic'] = any(0 < x < 1 for x in set(four_vector))
 
     def strategy(self, opponent):
         if not hasattr(self, "_four_vector"):
             raise ValueError("Fourvector not yet set")
-        if not len(opponent.history):
+        if len(opponent.history) == 0:
```
Comment (CamDavidsonPilon) [inline axelrod/strategies/memoryone.py]: Safer to checkout explicitly against 0, rather than using `not`

File: axelrod/strategies/memoryone.py [hunk 2]
```diff
@@ -106,7 +84,7 @@ def __init__(self, p=None):
 
     def receive_tournament_attributes(self):
         (R, P, S, T) = self.tournament_attributes["game"].RPST()
-        if not self.p:
+        if self.p is None:
```
Comment (CamDavidsonPilon) [inline axelrod/strategies/memoryone.py]: what if `p` is 0? This would previously fail the condition, but is a valid input. Instead I check against `None`

Comment (marcharper) [inline axelrod/strategies/memoryone.py]: Good catch!

File: axelrod/tests/unit/test_classification.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def test_multiple_instances(self):
         self.assertNotEqual(P1.classifier, P2.classifier)
 
         P1 = axelrod.GoByMajority(5)
-        P2 = axelrod.StochasticWSLS(10)
+        P2 = axelrod.StochasticWSLS(.1)
```
Comment (CamDavidsonPilon) [inline axelrod/tests/unit/test_classification.py]: This was failing because `1 - 10 < 0 < 1` and was caught in the ValueError added

**Comment by marcharper** [comment]
I'm fine with this. A few notes should you continue with this part of the library:

* There are several other memory one strategies not coded as special cases of `MemoryOnePlayer` -- some of the TFT players, of course Cooperator and Defector. We do like having some simple to follow strategies for newcomers, so let's not do the same to any of the MemoryZero strategies (Cooperator, Defector, and Random, at least).
* You could also make WSLS a special case of StochasticWinStayLoseShift. Just a thought. TFT could be a special case of GTFT as well and I am sure there are other examples in the library. It's ok to have multiple implementations I think -- it could be useful to test against each other when there is more than one implementation.
* I think that the previous code was (very) slightly more efficient (no random numbers). Perhaps we could modify `MemoryOnePlayer` to check for `p=0` or `p=1` and not draw a random in that case. I'm not sure if this would be a net gain or not, or if the difference is worth worrying about (probably not). Just a thought if you keep looking into this. Alternatively it might be better to leave the memory one players that have all probabilities equal to one or zero as is rather than as subclasses.

Anyone else have thoughts?

Some of the other maintainers are in Europe, so it may be a while before they can take a look (I think it's nighttime there now).

**Comment by CamDavidsonPilon** [comment]
@marcharper, thanks for reviewing. I agree that leaving some more popular models out of this, it just felt odd to see duplication in a single python file. 

w.r.t to efficiency, the random() call is very efficient. In fact, the `hasattr` call in the same function takes 10x as long as a call to `random`. So any perf hit would be negligible. 

I'll add a few tests to have coveralls pass

File: axelrod/strategies/memoryone.py (hunk 1 above)
Comment (drvinceknight) [inline axelrod/strategies/memoryone.py]: Can you break down why it's safer? What's the risk?

File: axelrod/strategies/memoryone.py (hunk 2 above)
Comment (drvinceknight) [inline axelrod/strategies/memoryone.py]: Reading over this briefly, could you add a test that checks that this fixes that potential case?

Comment (CamDavidsonPilon) [inline axelrod/strategies/memoryone.py]: yeye just added

**Comment by drvinceknight** [comment]
> There are several other memory one strategies not coded as special cases of MemoryOnePlayer -- some of the TFT players, of course Cooperator and Defector. We do like having some simple-to-follow strategies for newcomers, so let's not do the same to any of the MemoryZero strategies (Cooperator, Defector, and Random, at least).

I completely agree with this. Have just given a talk about the library and one of the things that was discussed was how nice it was that a lot of the strategies were very readable to non python experts.

> You could also make WSLS a special case of StochasticWinStayLoseShift. Just a thought. TFT could be a special case of GTFT as well and I am sure there are other examples in the library. It's ok to have multiple implementations I think -- it could be useful to test against each other when there is more than one implementation.

Not sure I'm a big fan of this suggestion as it's a contradiction with the previous one no? 

I'm also not sure about the multiple implementations idea (again: confusing for the non expert)... The fact that there is the generic memory one class means a user could code up other implementations if they wanted to but I'm not a fan of this being in the library... Perhaps I'm missing something?

Otherwise this all looks great, just the one suggestion about a test for what seems to be a bug that you caught?

File: axelrod/strategies/memoryone.py (hunk 2 above)
Comment (drvinceknight) [inline axelrod/strategies/memoryone.py]: :+1:

**Comment by CamDavidsonPilon** [comment]
Updated with more tests.

Conversely, one can see this refactor of `WinStayLoseShift` as an example implementation for the next non-expert to see how to implement a deterministic strategy using the `MemoryOnePlayer` class (which is super easy btw).

**Comment by drvinceknight** [comment]
> Conversely, one can see this refactor of WinStayLoseShift as an example implementation for the next non-expert to see how to implement a deterministic strategy using the MemoryOnePlayer class (which is super easy btw).

Very true but I would suggest your non expert there is more familiar with Python than my non expert...

There are quite a few strategies already present that show how to implement a MemoryOnePlayer... 

TitForTat as a verbose one liner I would not like to be changed. It was something I genuinely spoke about for a while during my talk...

**Comment by drvinceknight** [comment]
Thanks for the test: I'm happy with this PR as it is, perhaps we can open an issue to continue the discussion about TitForTat etc...?

**Comment by CamDavidsonPilon** [comment]
Oh no need to the issue, I agree with you two. My comment was that there was no _deterministic_ strategy that used `MemoryOnePlayer` (that I can see at least)

**Comment by marcharper** [comment]
I wasn't trying to suggest that TFT should be reimplemented as MemoryOne, I was just pointing out that there are a lot of cases where simply implemented strategies are special cases of parameterized strategies. (My wording was poor -- I meant the proverbial "you").

Maybe it's worth pointing these cases out in the docs for the relevant strategies, or in the docstring, and perhaps implementing tests for these corner cases.

**Comment by drvinceknight** [comment]
> Maybe it's worth pointing these cases out in the docs for the relevant strategies, or in the docstring, and perhaps implementing tests for these corner cases.

I really like that idea :)

**Comment by drvinceknight** [comment]
I think this can be merged or am I missing something?

**Comment by marcharper** [comment]
Sure, I'll update the "duplicate strategies" issue with the suggestion above.
